spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/serializer/TcpCodecs.java

```diff
@@ -93,21 +94,38 @@ public static ByteArrayLengthHeaderSerializer headerL1() {
 	/**
 	 * @return a {@link ByteArrayLengthHeaderSerializer} with a 2 byte header.
 	 */
-	public static ByteArrayLengthHeaderSerializer headerL2() {
+	public static ByteArrayLengthHeaderSerializer lengthHeader2() {
 		if (twoByteLHS == null) {
-			twoByteLHS = new ByteArrayLengthHeaderSerializer(1);
+			twoByteLHS = new ByteArrayLengthHeaderSerializer(2);
 		}
 		return twoByteLHS;
 	}
 
 	/**
 	 * @return a {@link ByteArrayLengthHeaderSerializer} with a 4 byte header.
 	 */
-	public static ByteArrayLengthHeaderSerializer headerL4() {
+	public static ByteArrayLengthHeaderSerializer lengthHeader4() {
 		if (fourByteLHS == null) {
-			fourByteLHS = new ByteArrayLengthHeaderSerializer(1);
+			fourByteLHS = new ByteArrayLengthHeaderSerializer(4);
 		}
 		return fourByteLHS;
 	}
 
+	/**
+	 * @param bytes header length.
+	 * @return a {@link ByteArrayLengthHeaderSerializer} with a 1, 2 or 4 byte header.
+	 */
+	public static ByteArrayLengthHeaderSerializer lengthHeader(int bytes) {
+		switch (bytes) {
+		case 1:
+			return lengthHeader1();
+		case 2:
+			return lengthHeader2();
+		case 4:
+			return lengthHeader4();
+		default:
+			throw new IllegalArgumentException("Only 1, 2 or 4 byte headers are supported");
```

Comment:
Oh! I see now:
```
public ByteArrayLengthHeaderSerializer(int headerSize) {
	if (headerSize != HEADER_SIZE_INT &&
		headerSize != HEADER_SIZE_UNSIGNED_BYTE &&
			headerSize != HEADER_SIZE_UNSIGNED_SHORT) {
		throw new IllegalArgumentException("Illegal header size:" + headerSize);
	}
	this.headerSize = headerSize;
}
```
So, we really can have only 3 possible variants.
Therefore I think your original solution without this `int bytes` variant was correct.

Although we can live with this one as well, e.g. for case of external configuration.
But let's really delegate assertion to the `ByteArrayLengthHeaderSerializer` ctor! 

When `ByteArrayLengthHeaderSerializer` will be changed, we may forget to change this factory method. Therefore - inconsistency. Although that might be just my paranoia and we can merge it as is :smile: 

WDYT?

BTW, are you going to reflect it in the docs somehow?